presto-thrift/src/main/java/com/facebook/presto/thrift/util/RetryDriver.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.thrift.util;
+
+import com.google.common.base.Throwables;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static java.util.Objects.requireNonNull;
+
+public class RetryDriver
```

Comment:
Moving it to a common project would be ideal. I can remove this one once it's there. However, there are slight modifications to this class I made to make it more suitable for this use case.